### DISCUSSION TITLE
OTEthernet not working

### DISCUSSION THREAD

**Discussion by wandering-bit** [JAndrassy/ArduinoOTA#204] [Q&A]
Components used:
1. Laptop
2. Offline home router
3. Wiznet w5500-evb-pico
The pico is connected to the router though the ethernet and the laptop is connected to the same router through the ethernet. After uploading the code to pico through com ports. The ip address of the pico is visible in the arduino ide network ports. However, when i try to upload a new code by selecting the network port. I get the below result. 
![Screenshot (20)](https://github.com/JAndrassy/ArduinoOTA/assets/72465815/be5729c6-e204-416b-8aea-cb6ec55f5da4)

**Comment by JAndrassy** [comment]
in IDE Preferences dialog turn on verbose mode for upload and try again. then copy the upload command and its output and post it here

**Reply by wandering-bit** [reply]
Sketch uses 68868 bytes (3%) of program storage space. Maximum is 1830912 bytes.
Global variables use 10528 bytes (4%) of dynamic memory, leaving 251616 bytes for local variables. Maximum is 262144 bytes.
"C:\Users\sranj\AppData\Local\Arduino15\packages\rp2040\tools\pqt-python3\1.0.1-base-3a57aed/python3" -I "C:\Users\sranj\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.2.0/tools/espota.py" -i "192.168.137.107" -p "65280" "--auth={upload.field.password}" -f "C:\Users\sranj\AppData\Local\Temp\arduino\sketches\79B6D195CDF9A7A6E317B16399E4997A/test1.ino.bin"
15:45:13 [ERROR]: No Answer
Failed uploading: uploading error: exit status 1

**Comment by JAndrassy** [comment]
it executes the default Pico W OTA upload tool. 
did you copy the platform.local.txt?

**Reply by wandering-bit** [reply]
to be double sure. the file
"C:\Users\sranj\Documents\Arduino\libraries\ArduinoOTA\extras\rp2040\platform.local.txt"
needs to be copied to
"C:\Users\sranj\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.2.0"
Am i doing it correctly?

**Reply by JAndrassy** [reply]
yes

**Reply by wandering-bit** [reply]
it is still not working. does it use epsota.py for uploading the code. is that the issue? 
Is it possible to execute the command through command line. If so, please let me know what should i type in.

**Reply by JAndrassy** [reply]
did you restart the IDE?

**Reply by wandering-bit** [reply]
yes, i restarted the ide, disabled the firewall as well. Still the same output

**Reply by JAndrassy** [reply]
take the platform.local.txt from this repo. it is updated for IDE 2

**Reply by wandering-bit** [reply]
still the same issue. it is still using espota.py

**Comment by JAndrassy** [comment]
you can download the fixed platform.local.txt

**Reply by ZinnerC** [reply]
@wandering-bit did you ever resolve your issue? I did not manage to solve at this point.
![image](https://github.com/JAndrassy/ArduinoOTA/assets/7002769/8f38481f-47aa-4590-95bc-74280c2eee6d)
This is where I placed the platform.local.txt file.
However I somehow get the error: "error: 'InternalStorage' was not declared in this scope"

**Reply by JAndrassy** [reply]
what board did you select in Tools menu?

**Reply by ZinnerC** [reply]
I selected once the "WIZnet W5100S-EVB-Pico" and once "Raspberry Pi Pico", both did not work.
![image](https://github.com/JAndrassy/ArduinoOTA/assets/7002769/84dd066d-1e7a-4cad-ae75-e9a59546e828)
I get both the same error message:
`In file included from C:\Users\CassianZinner\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\2.6.5\libraries\WiFi\src/WiFi.h:11,
                 from C:\Users\CassianZinner\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\2.6.5\libraries\ArduinoOTA\src/ArduinoOTA.h:24,
                 from C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA\examples\OTEthernet\OTEthernet.ino:22:
C:\Users\CassianZinner\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\2.6.5\libraries\WiFi\src/WiFiUdp.h:27: warning: "UDP_TX_PACKET_MAX_SIZE" redefined
   27 | #define UDP_TX_PACKET_MAX_SIZE 8192
      | 
In file included from C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA\examples\OTEthernet\OTEthernet.ino:21:
C:\Users\CassianZinner\AppData\Local\Arduino15\libraries\Ethernet\src/Ethernet.h:150: note: this is the location of the previous definition
  150 | #define UDP_TX_PACKET_MAX_SIZE 24
      | 
C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA\examples\OTEthernet\OTEthernet.ino: In function 'void setup()':
C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA\examples\OTEthernet\OTEthernet.ino:45:63: error: 'InternalStorage' was not declared in this scope
   45 |   ArduinoOTA.begin(Ethernet.localIP(), "Arduino", "password", InternalStorage);
      |                                                               ^~~~~~~~~~~~~~~
Multiple libraries were found for "ArduinoOTA.h"
  Used: C:\Users\CassianZinner\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\2.6.5\libraries\ArduinoOTA
  Not used: C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA
exit status 1

Compilation error: 'InternalStorage' was not declared in this scope`

and this is the code I use (which is the respective example).

`/*

 This example polls for sketch updates over Ethernet, sketches
 can be updated by selecting a network port from within
 the Arduino IDE: Tools -> Port -> Network Ports ...

 Circuit:
 * W5100, W5200 or W5500 Ethernet shield attached

 created 13 July 2010
 by dlf (Metodo2 srl)
 modified 31 May 2012
 by Tom Igoe
 modified 16 January 2017
 by Sandeep Mistry
 Ethernet version August 2018
 by Juraj Andrassy
 */
 
#include <SPI.h>
#include <Ethernet.h>
#include <ArduinoOTA.h>

//#define Serial SerialUSB

// Enter a MAC address for your controller below.
// Newer Ethernet shields have a MAC address printed on a sticker on the shield
byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };

void setup() {
  //Initialize serial:
  Serial.begin(9600);
  while (!Serial);

  // start the Ethernet connection:
  Serial.println("Initialize Ethernet with DHCP:");
  if (Ethernet.begin(mac) == 0) {
    Serial.println("Failed to configure Ethernet using DHCP");
  } else {
    Serial.print("  DHCP assigned IP ");
    Serial.println(Ethernet.localIP());
  }

  // start the OTEthernet library with internal (flash) based storage
  ArduinoOTA.begin(Ethernet.localIP(), "Arduino", "password", InternalStorage);
}

void loop() {
  // check for updates
  ArduinoOTA.handle();

  // add your normal loop code below ...
}
`

**Reply by JAndrassy** [reply]
did you delete the ArduinoOTA library bundled with the Pico core?

**Reply by ZinnerC** [reply]
I don‘t think so. Might it be that I have a version of the core were OTA is not yet possible?

**Reply by ZinnerC** [reply]
The screenshot where I place the platform.local.txt shows 2.6.5 but currently we have 3.2.1. Do I might require to update?

**Reply by JAndrassy** [reply]
from your output
```
Used: C:\Users\CassianZinner\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\2.6.5\libraries\ArduinoOTA
Not used: C:\Users\CassianZinner\Documents\Arduino\libraries\ArduinoOTA
```
the only way to suppress the bundled library is to delete it (it is a different library and only works with WiFi of Pico W)

**Reply by ZinnerC** [reply]
Thanks I did that and it compiled. However I don't get an IP over DHCP and when I try to use a fixed IP it tells me "(ip unset)" when I print Ethernet.localIP().

Could it be that the W5100s (emphesis on the S) is not supported? The example states "Circuit: * W5100, W5200 or W5500 Ethernet shield attached")

**Reply by JAndrassy** [reply]
did you set the right CS pin with Ethernet.init?

**Reply by ZinnerC** [reply]
Yes how I see it I did.

```C++
/*

 This example polls for sketch updates over Ethernet, sketches
 can be updated by selecting a network port from within
 the Arduino IDE: Tools -> Port -> Network Ports ...

 Circuit:
 * W5100, W5200 or W5500 Ethernet shield attached

 created 13 July 2010
 by dlf (Metodo2 srl)
 modified 31 May 2012
 by Tom Igoe
 modified 16 January 2017
 by Sandeep Mistry
 Ethernet version August 2018
 by Juraj Andrassy
 */
 
#include <SPI.h>
#include <Ethernet.h>
#include <ArduinoOTA.h>

#define W5100S_CS 17

//#define Serial SerialUSB

// Enter a MAC address for your controller below.
// Newer Ethernet shields have a MAC address printed on a sticker on the shield
byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
IPAddress ip(192,168,55,4);
IPAddress gateway(192,168,55,1);
IPAddress subnet(255,255,255,0);
IPAddress dns(8,8,8,8); 

void setup() {
  //Initialize serial:
  Serial.begin(9600);
  while (!Serial);

  // start the Ethernet connection:
  Serial.println("Initialize Ethernet with DHCP:");
  Ethernet.init(W5100S_CS);
  if (Ethernet.begin(mac) == 0) {
    Serial.println("  Failed to configure Ethernet using DHCP");

    Serial.println("  Try Static IP");
    Ethernet.begin(mac,ip);
    Serial.print("    Static assigned IP ");
    Serial.println(Ethernet.localIP());
  } else {
    Serial.print("  DHCP assigned IP ");
    Serial.println(Ethernet.localIP());
  }

  // start the OTEthernet library with internal (flash) based storage
  ArduinoOTA.begin(Ethernet.localIP(), "Arduino", "password", InternalStorage);
}

void loop() {
  // check for updates
  ArduinoOTA.poll();

  // add your normal loop code below ...
}
```

And the serial monitor gives me:
17:06:58.035 -> Initialize Ethernet with DHCP:
17:06:58.622 ->   Failed to configure Ethernet using DHCP
17:06:58.622 ->   Try Static IP
17:06:58.671 ->     Static assigned IP (IP unset)
